### PR TITLE
CORS-3485: Azure: modify control plane outbound LB into submission as an API LB

### PR DESCRIPTION
**Pull request by jhixson74**
https://issues.redhat.com/browse/CORS-3485

### CODE REVIEW

File: pkg/infrastructure/azure/loadbalancer.go [hunk 1]
```diff
@@ -65,13 +65,46 @@ func createPublicIP(ctx context.Context, in *lbInput) (*armnetwork.PublicIPAddre
...
+	}
+
+	// Get the existing frontend configuration and backend address pool and
+	// create an additional frontend configuration mand backend address
+	// pool. Use the newly created public IP address with the additional
+	// configuration so we can setup load balancing rules for the external
+	// API server.
+	frontendIPConfigs := extLB.Properties.FrontendIPConfigurations
+	if len(frontendIPConfigs) == 0 {
+		return nil, fmt.Errorf("could not get frontendIPConfigs for external LB %s", *extLB.Name)
```
Comment (sadasu) [inline pkg/infrastructure/azure/loadbalancer.go:88]: Even if len(frontendIPConfigs) == 0, do we still want to append additionalFrontendIPConfig to it?

File: pkg/infrastructure/azure/loadbalancer.go [hunk 2]
```diff
@@ -65,13 +65,46 @@ func createPublicIP(ctx context.Context, in *lbInput) (*armnetwork.PublicIPAddre
...
+		Properties: &armnetwork.FrontendIPConfigurationPropertiesFormat{
+			PrivateIPAllocationMethod: to.Ptr(armnetwork.IPAllocationMethodDynamic),
+			PublicIPAddress:           pip,
+		},
+	}
+	frontendIPConfigs = append(frontendIPConfigs, additionalFrontendIPConfig)
+
+	backendAddressPools := extLB.Properties.BackendAddressPools
+	if len(backendAddressPools) == 0 {
+		return nil, fmt.Errorf("could not get backendAddressPools for external LB %s", *extLB.Name)
```
Comment (sadasu) [inline pkg/infrastructure/azure/loadbalancer.go:101]: Afaik, we are not retying this operation (updateExternalLoadBalancer). So, if we return from here, we would have added a frontendIPConfig but no backendAddressPools. 

We could either 1. Add the new additionalBackendAddressPool even if len(backendAddressPools) == 0 OR 2. Add additionalFrontendIPConfig and additionalBackendAddressPool together after all error checking is complete.

**Comment by jhixson74** [comment]
I broke it.

**Comment by patrickdillon** [comment]
This is a nice way of cutting out the extra load balancer with minimal code changes. 

The primary risk in modifying the CAPI-created LB is that the changes could, in theory, be reconciled away. The testing looks good so it seems this is not happening, but we should be aware of the risk and perhaps firm it up a bit more. I don't think this should hold up our progress but I really do think it's worth considering.

/approve

Also, the fact that  the changes are not reconciled away could indicate bigger issues with how the LBs are handled in CAPZ. We saw in AWS that when LB creation failed, it would not be retried. Could be a similar issue here.

File: pkg/infrastructure/azure/network.go [hunk 3]
```diff
@@ -130,12 +148,13 @@ func createExternalLoadBalancer(ctx context.Context, pip *armnetwork.PublicIPAdd
...
 					},
 				},
+				OutboundRules: extLB.Properties.OutboundRules,
 			},
 			Tags: in.tags,
 		}, nil)
 
 	if err != nil {
-		return nil, fmt.Errorf("cannot create load balancer: %w", err)
+		return nil, fmt.Errorf("cannot update load balancer: %w", err)
```
Comment (sadasu) [inline pkg/infrastructure/azure/network.go:157]: Nit: cannot update external load balancer

**Comment by patrickdillon** [comment]
For user-defined routing, we're going to want to go back to creating the external LB rather than updating the control plane outbound LB. Based on this PR, it looks like that should be relatively straightforward...

**Pull request merged**